### PR TITLE
implement `ByteArrayInputStream.readNBytes()`

### PR DESCRIPTION
**Pull request by fiatjaf**
Hoping GitHub will test this for me automatically.

Closes https://github.com/scala-native/scala-native/issues/2774

### CODE REVIEW

**Comment by armanbilge** [comment]
Nice! You may want to target this PR at the 0.4.x branch if you want it in the next release :)

**Review by WojciechMazur — APPROVED** [review]

**Comment by LeeTibbert** [comment]
fiatjaf thank you for this first contribution. 

I make a small re-organization recommendation in item 2 below.  If implementing
that change would be onerous to you, I can do a follow-up PR once this PR is merged.

My motivation, after years of trying to keep javalib organized, is:
* Fixing untidiness before it enters the repository is the best approach.
* If now is not possible, fixing it __soon__ after saves us and future maintainers 
  headache down the road, when original contributors are long gone. 
  Lore evaporates quickly.

Wojciech I do not want to make a first contribution so complicated that it
puts off contributors.  I do have some review concerns:

1) These two methods first appear in Java 11.  Is it now acceptable
     to put Java 11 methods in SN javalib?  Did the guidance change,
     sorry if I missed the memo.
    
     I have spent years having my wrist slapped by the build itself
     whenever I tried to add non-Java8 methods but I may
     be behind the times.

2) It strongly looks to me that these methods should be default (implemented) 
    `InputStream.scala`.  If I read the Java 11 Spec correctly, `ByteArrayInputStream`
    inherits these methods from `InputStream.scala`.  I can check the Java 17 spec.
    Implementing them higher in the inheritance hierarchy follow the spec more
    closely and have the benefit that other classes which inherit from 'InputStream`
    would now have the methods "for free".  Because these are Java 11, there
    are currently no use cases in javalib which would need an `override`.

    On a hasty read, it looks like the current implementation would work in 
    `InputStream`. The kindly provided Test would continue to provide at
    least one class which exercised the default method and its inheritance.

**Comment by armanbilge** [comment]
> 2\. On a hasty read, it looks like the current implementation would work in
>         `InputStream`.

@LeeTibbert possibly, but see my comments in https://github.com/scala-native/scala-native/issues/2774#issuecomment-1206583639. The problem is that for an arbitrary `InputStream` the `readNBytes` method may be blocking (per JavaDoc), which we cannot support in a single-threaded runtime. It can however be implemented for the special case of `ByteArrayInputStream`.

File: unit-tests/shared/src/test/scala/javalib/io/ByteArrayInputStream.scala
Comment (LeeTibbert) [inline unit-tests/shared/src/test/scala/javalib/io/ByteArrayInputStream.scala:1]: This file name would be more consistent with the convention used for other file in `unit-test` if it were named `ByteArrayInputStreamTest.scala`.

Excellent to see a Test provided for introduced features.

File: javalib/src/main/scala/java/io/ByteArrayInputStream.scala
Comment (LeeTibbert) [inline javalib/src/main/scala/java/io/ByteArrayInputStream.scala:48]: By my reading of the Java 11 spec for this method, this
line is out-of-spec. I recommend deletion and let the
local read(3args) method handle the Exception.
 ``` IndexOutOfBoundsException - If off is negative, len is negative, or len is greater than b.length - off```

That would also allow this implementation becoming an `InputStream` default method.

**Comment by fiatjaf** [comment]
@LeeTibbert thank you for your comment.

I don't want to mess up with the code organization you have here, and I am not at all qualified to say if this is a good PR or not, I am merely doing it because @armanbilge suggested that I did when I was having trouble getting this method to work -- not knowing it was from javalib 11 at all and that the scala-native javalib was stuck at 8.

I do think it is ugly and weird to add it just for `ByteArrayInputStream` and not for `InputStream`, and even more ugly to not add the other methods that exist in javalib11 but just this one.

I will close this PR for now to not cause trouble, but I am looking forward to contributing more in the future if and when I can help better!

**Comment by armanbilge** [comment]
I don't think this PR should be closed at all :( it is a good contribution. @fiatjaf I'm sorry you felt like you are causing trouble, that is not at all the case. I hope you may reconsider and open it again.

The scala-native Javalib is _not_ stuck at Java 8. For example you can see in the test suite that there are tests for specifically JDK 11 and JDK 15 APIs:
https://github.com/scala-native/scala-native/tree/main/unit-tests/shared/src/test

Due to the constraints of Scala Native 0.4.x being single-threaded we cannot implement everything exactly as it's done in the JDK. If we can implement this method for `ByteArrayInputStream` (and we can!) we definitely should since that helps people today. We can then improve the implementation in Scala Native 0.5.x to more closely match the JDK when multithreading is supported.

**Comment by LeeTibbert** [comment]
@fiatjaf I suggest/recommend re-opening this PR and letting it advance.
It has value as both as 'the thing as it is' and as a first time contributor.

You have done nothing wrong, indeed something right.  

I find there can be a lot of collegial discussion during review.
The intent is to improve the product, not bring down the submitter.

I have spent a good number of hours in the past day or so
responding to a senior reviewer who, understandably, thought 
some comments could be improved and aligned in a manner
consistent with the rest of the project.  Grumble, Grumble,
about nits, I have work to get on with!  I have to admit
as I re-worked the code, several times, to respond to
the reviewer's concern, I managed to entirely eliminate
the code which was so complex that it lead to confusing
comments. Confusing & misaligned comments are
gone (if I manage to pass CI).

(Supportive) review lead to improved code.

**Comment by LeeTibbert** [comment]
>The scala-native Javalib is not stuck at Java 8. 
> For example you can see in the test suite that there are tests for specifically JDK 11 and JDK 15 APIs:
> https://github.com/scala-native/scala-native/tree/main/unit-tests/shared/src/test

Nearest I can figure out the current decision rule is: javalib can implement methods from JDKs > 8
but can use only Java 8 methods to do so (because SN does some link magic to link magic 
to link against a JDK 8)  I believe the current recommendation/requirement is that _any_ 
methods in javalib not use scalalib methods. There is a `scalaOps` which brings some
scala methods to java data.

I have never see the above written down and communicated by a person in authority.
What I am advocating for is such a description, nailed to  the cathedral door.

**Comment by LeeTibbert** [comment]
> Due to the constraints of Scala Native 0.4.x being single-threaded we cannot implement everything 
> exactly as it's done in the JDK. 

Agreed, wholeheartedly. 

>If we can implement this method for ByteArrayInputStream (and we can!) we definitely should since
>  that helps people today. 
I concur, helping actual people is the first goal.  Theory guides action, action informs theory.

From a software engineering standpoint, I suggest that if necessity drives something  be done 
in a way a  reasonable person would not expect from the relevant Java official spec, then it
deserves a one line comment in the code referring to an Issue number, and a two line
entry in the PR saying, in better words "Foo is implemented different that the spec. See Issue #xxx".

The latter gives the person/people doing the merging that there is particular issue needs close attention.
The latter informs a line-by-line reviewer, and people downstream who have to maintain, why a 
particular design decision was made in an apparently astonishing way.

**Comment by LeeTibbert** [comment]
fiatjaf, not to distract or dissuade you by differing opinions.
As mentioned, I recommend you re-open this and let it flow
through.  That gives you a simple & successful walkthrough
(and something you and others can use in the near future).

That way you and your friends can get on to implementing 
other Javalib methods ;-) 

Anyone motivated to make changes can submit a PR later.
The Open Source process.

**Comment by LeeTibbert** [comment]
Arman,  as collegial discussion.    re: requirement for multithreading support

I must be missing something about the expressed requirement for multithread
support blocking implementing this from being in `InputStream`.  I am 
pretty sure that implementing there introduces no constraints that 
do not already exist today: if you read from a stream that can block, 
your process may block.  

If one is concerned about blocking, one must either not read from
a blocking stream or implement timeout code.

Alternately one can create a new class which inherits from the blocking stream, 
and overrides various read methods with (pseudo) non-blocking reads.

Consider `FileInputStream` today.  It has a method, which is implemented
in Scala Native. `int read(byte[] b, int off, int len)`, which can block. This is analogous
to the proposed InputStream.readNBytes. Agreed, a user of either can block,
especially if `isAvailable()` is unreliable.     How is the latter introducing badness
that is not already in the former?  

Implementing in `InputStream`, as per the spec, does not fix anything, but it
also does not introduce harm.

What am I missing?  How will multithreading help a thread blocked on I/O?

I will have to think about it some more but I think even POSIX Pthread 
would have problems reading data that never comes, say from dropped
messages from a  UDP protocol where the sender never re-sends
or the Consumer of the tail of a log file where the Producer dies without
warning or cleanup.

I agree entirely that I/O with thread support is much easier and more enjoyable.
--------------------------------------------------------------------------------------------------------------------------
There may be a need to look even more carefully at the EOF processing of
this PR to make sure it is handled correctly.

I would have to re-study the spec with EOF in mind. The first read
when the `mark` is at the end of the buffer may need to return 0 (EOF)
and subsequent throw the exception.

**Comment by LeeTibbert** [comment]
iatjaf  Thank you for re-opening and continuing the PR process.

Putting concerns about `InputStream` on the stack for awhile,
my only pending request is that you change the name of the 
unit-test file from 'foo' to 'fooTest'.

**Comment by armanbilge** [comment]
> Consider `FileInputStream` today. It has a method, which is implemented
> in Scala Native. `int read(byte[] b, int off, int len)`, which can block. This is analogous
> to the proposed InputStream.readNBytes

Thanks, great point. It is not exactly analogous. `readNBytes` will block until `N` bytes are available, whereas the `read` method may bail out early before `len` bytes are read, so it may not have to block.

But you are right. I reviewed the JavaDoc for `read` more carefully:

> This method blocks until input data is available, ... there is an attempt to read at least one byte.

Indeed it will always attempt to read at least one byte, and therefore it is not a valid implementation to bail out without reading any bytes so as to avoid blocking. This means implementing a non-blocking `InputStream` for an underlying blocking source is simply not possible.



> How is the latter introducing badness
> that is not already in the former?

Well put, I agree.

File: javalib/src/main/scala/java/io/ByteArrayInputStream.scala
Comment (LeeTibbert) [inline javalib/src/main/scala/java/io/ByteArrayInputStream.scala:47]: The squirrels running around inside my brain woke me out
of my nap to ask:  What happens if `len` in line 47 is negative?
Does the proper exception `IllegalArgumentException - if length is negative` get thrown?

I said "Don't be silly, no one would actually pass a negative
value. If they did, this code would throw `java.lang.NegativeArraySizeException` and they would be hosed."  Then I stretched, wrote this note, and went off to seek supper.

PS:  I believe that the 3arg variant is OK in this regard. It
        passes the negative value down to a read, which 
        then throws the `IndexOutOfBoundsException`  
        documented for that variant.

**Comment by LeeTibbert** [comment]
To summarize, I believe the consensus goal is to ship this PR
and let motivated people do follow-ons?


I think there are two tiny changes which are necessary. 

* The name of the test file should be `*Test.scala`
   (Lee why are you complaining about a file name? You
    should be rejoicing that there is a test file at all.)

* There is a test for negative len which would be
   really nice to handle here instead of a follow-on.

@fiatjaf  If you feel like making these changes, the 
file name change should be a `git mv`, `git commit`,
`git push` cycle.  If it causes you problems, yell
(to, not at, me).  That change is a _lot_ easier
if it happens in this PR, before the file gets
created in the repository.

If you are tired of faffing with this PR, I can 
do the len test in a, hopeful, follow-on for `InputStream`.

And no, your dog is not ugly...

**Comment by fiatjaf** [comment]
Sorry for the delay. I've pushed the 3 changes requested (or rather, mentioned) by @LeeTibbert:
1. Do not check for `n` bigger than `buf`
2. Check if `len` is negative and throw `IllegalArgumentException`
3. Change the name of the test file

**Comment by LeeTibbert** [comment]
> Sorry for the delay.  

Thank you for the timely changes. This is the weekend,
so that is fast turnaround in my book.
 
I appreciate all the changes, but especially the file name change.

**Comment by WojciechMazur** [comment]
Thank you for this PR, overall it looks good. There's a typo leading to compilation errors. Also, it's required to sign CLA at  https://www.lightbend.com/contribute/cla/scala - it's one of the reasons why the CI fails

**Comment by WojciechMazur** [comment]
Both of the added tests fail compliance tests with the JDK11 behavior. Probably due to the usage of assertEquals which for Arrays checks only their reference equality, not their content. 

From what I've seen in the spec readNBytes methods are defined in the `InputStream` and one of them is overridden in ByteArrayInputStream. We should move implementation to InputStream to automatically cover other types of InputStreams. 

Tests should be moved to `unit-tests/shared/src/test/require-jdk11`. Current tests cannot be compiled using JDK8. Also their package should be changed to `org/scalanative/testsutie/javalib/io` to match the current structure, and the test name should be appended with `onJDK11` to prevent name clashes in the future. 

It would be great to also provide `skipNBytes` method which was added in the same version of JDK for partity

**Comment by fiatjaf** [comment]
Thank you for the feedback @WojciechMazur.

I agree with you and that is why I had closed this PR in the first place. I'm not used enough to the codebase or Scala or Java to get these things correctly and I can't put time on it right now, although I wish I could. I'll close this again and hope someone else (or perhaps a future me) makes a better job later.

**Pull request closed without merging**
